File: solutions/java/src/main/java/com/epi/TeamPhoto2.java
package com.epi;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * @author translated from c++ by Blazheev Alexander
 */
public class TeamPhoto2 {
  private static class Player<HeightType extends Comparable<HeightType>>
      implements Comparable<Player<HeightType>> {
    public HeightType height;

    public Player(HeightType height) { this.height = height; }

    @Override
    public int compareTo(Player<HeightType> o) {
      return height.compareTo(o.height);
    }
  }

  private static class Team<HeightType extends Comparable<HeightType>> {
    private ArrayList<Player<HeightType>> members;

    public Team(List<HeightType> height) {
      members = new ArrayList<>();
      for (HeightType h : height) {
        members.add(new Player<>(h));
      }
    }

    public boolean lessThen(Team<HeightType> that) {
      List<Player<HeightType>> thisSorted = sortHeightMembers();
      List<Player<HeightType>> thatSorted = that.sortHeightMembers();
      for (int i = 0; i < thisSorted.size() && i < thatSorted.size(); ++i) {
        if (!(thisSorted.get(i).compareTo(thatSorted.get(i)) < 0)) {
          return false;
        }
      }
      return true;
    }

    private List<Player<HeightType>> sortHeightMembers() {
      List<Player<HeightType>> sortedMembers =
          (List<Player<HeightType>>)members.clone();
      Collections.sort(sortedMembers);
      return sortedMembers;
    }
  }

  // @include
  public static class GraphVertex {
    public List<GraphVertex> edges = new ArrayList<>();
    public int maxDistance = 1;
    public boolean visited = false;
  }

  public static int findLargestNumberTeams(List<GraphVertex> G) {
    LinkedList<GraphVertex> vertexOrder = buildTopologicalOrdering(G);
    return findLongestPath(vertexOrder);
  }

  private static LinkedList<GraphVertex> buildTopologicalOrdering(
      List<GraphVertex> G) {
    LinkedList<GraphVertex> vertexOrder = new LinkedList<>();
    for (GraphVertex g : G) {
      if (!g.visited) {
        DFS(g, vertexOrder);
      }
    }
    return vertexOrder;
  }

  private static int findLongestPath(LinkedList<GraphVertex> vertexOrder) {
    int maxDistance = 0;
    while (!vertexOrder.isEmpty()) {
      GraphVertex u = vertexOrder.peek();
      maxDistance = Math.max(maxDistance, u.maxDistance);
      for (GraphVertex v : u.edges) {
        v.maxDistance = Math.max(v.maxDistance, u.maxDistance + 1);
      }
      vertexOrder.pop();
    }
    return maxDistance;
  }

  private static void DFS(GraphVertex cur, LinkedList<GraphVertex> vertexOrder) {
    cur.visited = true;
    for (GraphVertex next : cur.edges) {
      if (!next.visited) {
        DFS(next, vertexOrder);
      }
    }
    vertexOrder.push(cur);
  }
  // @exclude

  public static void main(String[] args) {
    List<GraphVertex> G = new ArrayList<>(3);
    for (int i = 0; i < 3; i++) {
      G.add(new GraphVertex());
    }
    G.get(0).edges.add(G.get(2));
    G.get(1).edges.add(G.get(2));
    assert(2 == findLargestNumberTeams(G));
  }
}

File: static/solutions/Team_photo_2.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <limits>
#include <stack>
#include <vector>

using std::max;
using std::stack;
using std::vector;

struct GraphVertex;
stack<GraphVertex*> BuildTopologicalOrdering(vector<GraphVertex>* G);
int FindLongestPath(stack<GraphVertex*>* vertex_order);
void DFS(GraphVertex* cur, stack<GraphVertex*>* vertex_order);

// @include
struct GraphVertex {
  vector<GraphVertex*> edges;
  int max_distance = 1;
  bool visited = false;
};

int FindLargestNumberTeams(vector<GraphVertex>* G) {
  stack<GraphVertex*> vertex_order(BuildTopologicalOrdering(G));
  return FindLongestPath(&vertex_order);
}

stack<GraphVertex*> BuildTopologicalOrdering(vector<GraphVertex>* G) {
  stack<GraphVertex*> vertex_order;
  for (auto& g : *G) {
    if (!g.visited) {
      DFS(&g, &vertex_order);
    }
  }
  return vertex_order;
}

int FindLongestPath(stack<GraphVertex*>* vertex_order) {
  int max_distance = 0;
  while (!vertex_order->empty()) {
    GraphVertex* u = vertex_order->top();
    max_distance = max(max_distance, u->max_distance);
    for (GraphVertex*& v : u->edges) {
      v->max_distance = max(v->max_distance, u->max_distance + 1);
    }
    vertex_order->pop();
  }
  return max_distance;
}

void DFS(GraphVertex* cur, stack<GraphVertex*>* vertex_order) {
  cur->visited = true;
  for (GraphVertex* next : cur->edges) {
    if (!next->visited) {
      DFS(next, vertex_order);
    }
  }
  vertex_order->emplace(cur);
}
// @exclude

int main(int argc, char* argv[]) {
  vector<GraphVertex> G(3);
  G[0].edges.emplace_back(&G[2]);
  G[1].edges.emplace_back(&G[2]);
  assert(2 == FindLargestNumberTeams(&G));
  return 0;
}
